kura/org.eclipse.kura.driver.ble.sensortag.provider/src/main/java/org/eclipse/kura/internal/driver/ble/sensortag/SensorTagDriver.java

```diff
@@ -0,0 +1,500 @@
+/**
+ * Copyright (c) 2017 Eurotech and/or its affiliates and others
+ *
+ *  All rights reserved. This program and the accompanying materials
+ *  are made available under the terms of the Eclipse Public License v1.0
+ *  which accompanies this distribution, and is available at
+ *  http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Eurotech
+ */
+package org.eclipse.kura.internal.driver.ble.sensortag;
+
+import static java.util.Objects.isNull;
+import static java.util.Objects.requireNonNull;
+import static org.eclipse.kura.KuraErrorCode.OPERATION_NOT_SUPPORTED;
+import static org.eclipse.kura.channel.ChannelFlag.FAILURE;
+import static org.eclipse.kura.channel.ChannelFlag.SUCCESS;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+import org.eclipse.kura.KuraException;
+import org.eclipse.kura.KuraRuntimeException;
+import org.eclipse.kura.bluetooth.le.BluetoothLeAdapter;
+import org.eclipse.kura.bluetooth.le.BluetoothLeDevice;
+import org.eclipse.kura.bluetooth.le.BluetoothLeService;
+import org.eclipse.kura.channel.ChannelFlag;
+import org.eclipse.kura.channel.ChannelRecord;
+import org.eclipse.kura.channel.ChannelStatus;
+import org.eclipse.kura.channel.listener.ChannelListener;
+import org.eclipse.kura.configuration.ConfigurableComponent;
+import org.eclipse.kura.driver.ChannelDescriptor;
+import org.eclipse.kura.driver.Driver;
+import org.eclipse.kura.driver.PreparedRead;
+import org.eclipse.kura.driver.ble.sensortag.localization.SensorTagMessages;
+import org.eclipse.kura.localization.LocalizationAdapter;
+import org.eclipse.kura.type.DataType;
+import org.eclipse.kura.type.TypedValue;
+import org.eclipse.kura.type.TypedValues;
+import org.eclipse.kura.util.base.TypeUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The Class {@link SensorTagDriver} is a BLE SensorTag Driver implementation for
+ * Kura Asset-Driver Topology.
+ * <br/>
+ * <br/>
+ * This BLE SensorTag Driver can be used in cooperation with Kura Asset Model and in
+ * isolation as well. In case of isolation, the properties needs to be provided
+ * externally.
+ * <br/>
+ * <br/>
+ * The required properties are enlisted in {@link SensorTagChannelDescriptor} and
+ * the driver connection specific properties are enlisted in
+ * {@link SensorTagOptions}
+ *
+ * @see Driver
+ * @see SensorTagOptions
+ * @see SensorTagChannelDescriptor
+ *
+ */
+public final class SensorTagDriver implements Driver, ConfigurableComponent {
+
+    private static final Logger logger = LoggerFactory.getLogger(SensorTagDriver.class);
+    private static final SensorTagMessages message = LocalizationAdapter.adapt(SensorTagMessages.class);
+
+    private static final String DISCOVERY_STOP_EX = "Failed to stop discovery";
```

Comment:
This maybe can be exported in the localization.

Reply:
👍